### PR TITLE
Option to have http.server listen on a unix socket instead of TCP

### PR DESCRIPTION
**Pull request by ChickenNuggers**
Review (http://stackoverflow.com/questions/15716302/so-reuseaddr-and-af-unix) on whether SO_REUSEADDR should be kept?

### CODE REVIEW

**Comment by daurnimator** [comment]
- `cs.listen` should only be called in a single location.
  - might as well pass all the arguments (i.e. `v6only` and `reuseport`) that the user passes in. If they use a `path` and `reuseport` that's their problem. There won't be any problems if you pass in `host`,`port` and `path` all in the one constructor.
  - I'd prefer you didn't use the `host` variable to store a path, it's just confusing.
  - You need to write tests
  - Need to update docs

**Comment by daurnimator** [comment]
Furthermore, if you're allowing binding unix sockets, you should probably add support for the `cs.listen` options:
  - `.mode` string:nil fchmod or chmod socket after creating UNIX domain socket
  - `.mask` string:nil set and restore umask when binding UNIX domain sockets
  - `.unlink` boolean:false unlink socket path before binding

File: http/server.lua [hunk 1]
```diff
@@ -176,9 +177,11 @@ Takes a table of options:
 ]]
 local function listen(tbl)
 	local tls = tbl.tls
-	local host = assert(tbl.host, "need host")
+	local host = tbl.host
+	local path = tbl.path
+	assert(host or path, "need host or path")
 	local port = tbl.port
-	if port == nil then
+	if port == nil and tbl.host then
```
Comment (daurnimator) [inline http/server.lua]: Just use `host` instead of indexing again.

File: http/server.lua [hunk 2]
```diff
@@ -189,11 +192,12 @@ local function listen(tbl)
 	end
 	local ctx = tbl.ctx
 	if ctx == nil and tls ~= false then
```
Comment (daurnimator) [inline http/server.lua]: Perhaps error if `tls == false` and using a unix socket? I don't think there is a sensible default for the canonical name in the certificate.

File: spec/server_spec.lua [hunk 3]
```diff
@@ -50,16 +64,32 @@ describe("http.server module", function()
...
+	it("works with https 1.1 using UNIX socket domain", function()
+		simple_test(true, 1.1, os.tmpname() .. ".socket")
+	end)
+	it("works with plain http 2.0 using UNIX socket domain", function()
+		simple_test(false, 2.0, os.tmpname() .. ".socket")
+	end);
+	(require "http.tls".has_alpn and it or pending)("works with https 2.0 using UNIX socket domain", function()
+		simple_test(true, 2.0, os.tmpname() .. ".socket")
+	end)
+	for k, v in pairs(socket_paths) do
```
Comment (daurnimator) [inline spec/server_spec.lua]: I'd prefer you deleted *in* the test.

File: spec/server_spec.lua [hunk 4]
```diff
@@ -50,16 +64,32 @@ describe("http.server module", function()
...
 	end)
-	it("works with plain http 2.0", function()
+	it("works with plain http 2.0 using IP", function()
 		simple_test(false, 2.0)
 	end);
-	(require "http.tls".has_alpn and it or pending)("works with https 2.0", function()
+	(require "http.tls".has_alpn and it or pending)("works with https 2.0 using IP", function()
 		simple_test(true, 2.0)
 	end)
+	local socket_path = os.tmpname() .. ".sock"
```
Comment (daurnimator) [inline spec/server_spec.lua]: unused variable

File: spec/server_spec.lua [hunk 5]
```diff
@@ -1,5 +1,5 @@
 local TEST_TIMEOUT = 2
-describe("http.server module", function()
+describe("http.server module using hostnames", function()
```
Comment (daurnimator) [inline spec/server_spec.lua]: why the new describe string?

**Comment by daurnimator** [comment]
Blocked on https://github.com/daurnimator/lua-http/issues/9

**Comment by daurnimator** [comment]
#9 was resolved.

Are you able to resolve my other comments?

File: spec/server_spec.lua [hunk 6]
```diff
@@ -13,30 +13,46 @@ describe("http.server module", function()
...
+		local host, port
+		if not path then
+			local _
+			_, host, port = s:localname()
+		end
 		local on_stream = spy.new(function(stream)
 			stream:get_headers()
 			stream:shutdown()
 			s:shutdown()
+			print 'shutting down server // spy'
```
Comment (daurnimator) [inline spec/server_spec.lua]: please remove these prints.

File: http/server.lua (hunk 2 above)
Comment (ChickenNuggers) [inline http/server.lua]: I'm confused on what you mean by this.

Comment (daurnimator) [inline http/server.lua]: I'm saying something like (feel free to write a nicer error message):
```
if ctx == nil and tls ~= false then
    if host then
        ctx = new_ctx(host)
    else
        error("when using a UNIX socket you must pass a TLS context if you want to use TLS")
    end
end
```

Comment (ChickenNuggers) [inline http/server.lua]: Fixed but currently the spec is `pending()` until contexts can be generated.

**Comment by ChickenNuggers** [comment]
Closing for reformatting commits, will create a new PR later.

**Pull request closed without merging**